PR: Add font properties handling

goanpeca opened the pull request.

freakboy3742 commented on colosseum/validators.py:
```diff
@@ -103,3 +106,39 @@ def is_color(value):
+            checked_values.append(val)
+        else:
+            # TODO: Check that the font exists?
+            if _CSS_IDENTIFIER_RE.match(val):
```
Is there a CSS spec reference on this character restriction?

goanpeca replied:
https://www.w3.org/TR/CSS22/fonts.html#font-family-prop

>Font family names must either be given quoted as strings, or unquoted as a sequence of one or more identifiers. This means most punctuation characters and digits at the start of each token must be escaped in unquoted font family names.

>For example, the following declarations are invalid:

[On identifiers](https://www.w3.org/TR/CSS22/syndata.html#value-def-identifier)

freakboy3742 replied:
Ok - in which case, this (or a section reference) should be in the code as a docstring (probably around the regex declaration, rather than the usage). Essentially, I'm aiming for this to be as close to "literate code" as we can manage - any policy decision in code should be referencing the standard so that we can easily identify which decisions are arbitrary implementation decisions, and which are required by the standard.

The examples of invalid identifiers should also be included as test cases.

freakboy3742 commented on colosseum/declaration.py:
```diff
@@ -276,22 +317,22 @@ def __init__(self, **style):
     # 15. Fonts ##########################################################
     # 15.3 Font family
-    # font_family
+    font_family = validated_property('font_family', choices=FONT_FAMILY_CHOICES, initial=INITIAL)  # TODO: initial?
```
font_family isn't a "normal" property, because it can be multivalued. The single string representation is fine for output/debug purposes, but end consumers of a style really need the list of pre-parsed font names, rather than a string that needs to be repeatedly parsed.

goanpeca replied:
Will fix!

goanpeca replied:
@freakboy3742 would it be better to make this a tuple? or how should we handle mutability here, if the user pops a value from the list? Should we return a copy of the list?

goanpeca replied:
@freakboy3742 what do you mean by 

>this isn't a tuple - position isn't significant, only ordering.

freakboy3742 replied:
This is one of those theoretical aspects of API design that isn't always followed well because it's not obvious that it exists. The underlying question - why is there two "ordered" datatypes in Python?

Consider the coordinates for a point, and a path made up of a collection of points.

A point is a *tuple* of three values - x, y and z. It must *always* be three values - you can't add a new value; and there's significance to each position - the second value *is* the coordinate on the y axis.

A path is a *list*. The list can be of arbitrary length, and you can add new values to the end of the list. However, there's no particular significance to the each position, other than the ordering - there's nothing especially significant about position 2, other than the fact that it comes after position 1, and before position 3 (if it exists).

There's one other notable difference: every value in a list *should* be the same data type (or, at least, have some common duck-base). There's no such requirement of a tuple. In the case of point, they all happen to be coordinates - but consider something like an RGBA color specification. RGB are integers/hex values; A is a percentage. 

A tuple isn't just an immutable list. Immutability is a useful property when all the positions are significant, but it's not the most interesting distinguishing property between a List and a Tuple. 

Put another way - If Python were to remove index-based addressing from lists, it should be *inconvenient*, but not problematic, because lists should be iterated over. Tuples, on the other hand, aren't generally interesting to iterate over, because you need to handle each position differently. In practice, this isn't *completely* true, because indexing lists can be useful when iterating, and good functional programming means you can use positional based function dispatch, but hopefully my underlying point is clear.

goanpeca replied:
Perfectly clear now :-), thanks for the thorough explanation

freakboy3742 commented on tests/test_validators.py:
```diff
@@ -38,3 +39,23 @@ def test_number(self):
+        ]
+        for case in invalid_cases:
+            with self.assertRaises(ValidationError):
+                validator(case)
```
As a stylist thing - why rename is_font_family to validator for this use case?

goanpeca replied:
Hmmmm I had two separate validators and then turned them into one, this is a leftover of that. Will fix

goanpeca commented on colosseum/validators.py:
```diff
@@ -103,3 +106,39 @@ def is_color(value):
+    value = ' '.join(value.strip().split())
+    values = [v.strip() for v in value.split(',')]
+    checked_values = []
+    generic_family = ['serif', 'sans-serif', 'cursive', 'fantasy', 'monospace']
```
👍

goanpeca commented on colosseum/constants.py:
```diff
@@ -317,26 +318,107 @@ def __str__(self):
+    'font_weight': NORMAL,
+    'font_size': MEDIUM,
+    'line_height': NORMAL,
+    'font_family': '',  # TODO: Depends on user agent. What to use?
```
Ok

goanpeca replied:
So... `SYSTEM = 'system'` or `INITIAL` for now ?

goanpeca commented on colosseum/declaration.py:
```diff
@@ -12,12 +12,53 @@
+            # Attribute doesn't exist
+            pass
+
+        # TODO: Should this delete all the other attributes?
```
Ok

goanpeca commented on colosseum/constants.py:
```diff
@@ -317,26 +318,107 @@ def __str__(self):
+    'font_variant': NORMAL,
+    'font_weight': NORMAL,
+    'font_size': MEDIUM,
+    'line_height': NORMAL,
```
Nope, I need to define it ! yu are right

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,93 @@
+
+# Find spaces within single and double quotes
+SQ_PATTERN = re.compile(r"\s+(?=(?:(?:[^']*'){2})*[^']*'[^']*$)")
+DQ_PATTERN = re.compile(r'\s+(?=(?:(?:[^"]*"){2})*[^"]*"[^"]*$)')
```
😆

goanpeca replied:
Open for suggestions, this was the "simplest" I could come up with.

goanpeca replied:
>Although it's a very low-fi solution, I think my preference here would be to do a very simple tokenizer - i.e., something that iterates over a string one character at a time, yielding a "word" when a separator is seen, unless a quote character is seen, in which case you enter in-a-quote state until a second quote is seen. We can then back that with a thorough suite of tests (that we can add to as we find pathological edge cases), and have an easily inspectable understanding of what is/isn't valid.

Cool, I have done this in the past. Will add a simple tokenizer then :-)

goanpeca replied:
So this was fixed with your other suggestion

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,93 @@
+    # Replace commas between font families with some character
+    string = string.replace(comma, comma_sep)
+    string = string.replace(comma[0], comma_sep)
+    string = replace_font_family_spaces(string, space_sep)
```
Since all these are options, and there could be several font families and some font families have quotes (if using weird characters or using spaces) I thought the easiest would be to replace the comma and space there and the spaces within quotes and then do a simple split.

```
font: 1.2em "Fira Sans", sans-serif;
font: italic 1.2em "Fira Sans", serif;
font: italic small-caps bold 16px/2 cursive;
font: small-caps bold 24px/1 sans-serif;
font: caption;
```

>Since we're already knee deep with a complex regex

I would say it does not help much since the regex would end up being uber complex (AFAICT)

goanpeca replied:
I am open for suggestions here :-)

freakboy3742 replied:
Ok - but I think we can work the underlying structure of the font specifier to our advantage here.

The font family must be the the last "element" in the specifier; everything before that point is either from a "known keyword" list (`bold`/`italic`/...) or fits a known pattern (`\dpx`,...). My approach would be to `lsplit(' ', 1)` from the front of the specifier trying to match the "known" patterns until you find a split that won't work - at which point, you take the version of the string prior to the last `lsplit`, and try and parse it as a font family.

goanpeca replied:
That could work as well :-)

goanpeca replied:
Ok, added the suggested fix

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            if value not in SYSTEM_FONT_KEYWORDS:
+                # TODO: Should this be a different error, e.g. ParsingError?
+                raise ValidationError
```
Ok

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,93 @@
+        if '/' in font['font_size']:
+            font['font_size'], font['line_height'] = font['font_size'].split('/')
+    else:
+        # TODO: Should this be a different error, e.g. ParsingError?
```
Ok

goanpeca commented:
Another question: Should the setter of `font` receive a tuple of values where the last item might be a list? or have the `setter(self, *args)` (does this even work for setters :-p ?) ? Should we have several options?


```python

# Which one should we provide, enforce?

node.style.font = 'normal', 'normal', 'normal', '12px/10px',  'palatino'
node.style.font = ('normal', 'normal', 'normal', '12px/10px',  'palatino')
node.style.font = ['normal', 'normal', 'normal', '12px/10px',  'palatino']
node.style.font = ['normal', 'normal', 'normal', '12px/10px',  'palatino, serif']
node.style.font = ['normal', 'normal', 'normal', '12px/10px',  ['palatino, serif']]
```

goanpeca commented:
Hi @freakboy3742 made the fixes as suggested and simplify the parsing (no more regex! yei).

Added a new list validated property and completed tests for font_family and font shorthand. Left a couple of questions as `TODO:`

I could add validation of system fonts by querying for available fonts or I could leave that for a separate PR

goanpeca commented on colosseum/constants.py:
```diff
@@ -317,26 +323,161 @@ def __str__(self):
+
+def available_font_families():
+    """List available font family names."""
+    # TODO: for tests
```
I should probably mock this for tests 😆

freakboy3742 commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,95 @@
+            except ValueError:
+                pass
+
+        raise ValidationError
```
Should raise an instance, not the class.

goanpeca replied:
Fixed

freakboy3742 commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,95 @@
+
+        # Maybe it is a font size
+        if '/' in value:
+            font['font_size'], font['line_height'] = value.split('/')
```
This won't validate that it's a valid size and/or line height - any '/' separated value will be accepted.

goanpeca replied:
Fixed

freakboy3742 commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,95 @@
+                raise ValidationError(error_msg)
+            font = get_system_font(value)
+    else:
+        for _ in range(5):
```
It would be worth dropping a quick explanatory note here; I suggested this technique, and it still took me a moment to decode what you were doing with this approach :-)

freakboy3742 commented on tests/test_declaration.py:
```diff
@@ -657,3 +689,53 @@ def test_dict(self):
+        node.style.line_height = '1.5'
+        node.style.font_family = ['"DejaVu Sans"', 'serif']
+        # TODO: Is this the behavior we want?
+        self.assertEqual(node.style.font, 'italic small-caps bold 10.0px/1.5 "DejaVu Sans", serif')
```
I'd say this is definitely what `__str__` should return on a font; but the font itself should probably be a wrapper object of some kind.

goanpeca replied:
@freakboy3742 would it be ok to just use a dictionary? (as the one we use for the initial font values?)

freakboy3742 commented on tests/test_declaration.py:
```diff
@@ -657,3 +689,53 @@ def test_dict(self):
+        node.style.font_size = '10px'
+        node.style.line_height = '1.5'
+        node.style.font_family = ['"DejaVu Sans"', 'serif']
+        self.assertEqual(node.style.font, '9px serif')
```
This isn't the behavior I'd expect (especially if they're set in this order) - I'd expect the final font str to be `italic small-caps bold 10px/1.5 "DejaVu Sans", serif`. This matches what you'd expect if the calls were made sequentially (and also matches CSS behavior if the declarations are ordered in the same order)

goanpeca replied:
Got it, was not sure of the behavior and did not test it myself

freakboy3742 commented on tests/test_declaration.py:
```diff
@@ -657,3 +689,53 @@ def test_dict(self):
+
+        # Check invalid values
+        with self.assertRaises(ValueError):
+            node.style.font = 'foobar'
```
May need to be careful here - there actually is a font called "foobar" (which I happen to have installed :-). If we're picking something that will be interpreted as a font name, we need to be certain it won't match - something like `"ThisIsDefinitelyNotAFontName"`

goanpeca replied:
😆 ok, fixed it!

goanpeca commented on colosseum/declaration.py:
```diff
@@ -12,22 +12,80 @@
 
     def getter(self):
-        return getattr(self, '_%s' % name, initial)
+        font = initial
```
Yep, it should!

goanpeca commented on colosseum/constants.py:
```diff
@@ -317,26 +323,174 @@ def __str__(self):
+
+AVAILABLE_FONT_FAMILIES = available_font_families()
+FONT_FAMILY_CHOICES = Choices(
+    validators=[is_font_family(generic_family=GENERIC_FAMILY_FONTS, font_families=AVAILABLE_FONT_FAMILIES)],
```
Fixed!

goanpeca commented on .github/workflows/ci.yml:
```diff
@@ -30,6 +30,10 @@ jobs:
+    - name: Copy Ahem font
+      run: |
+        mkdir -p ~/.local/share/fonts/
+        cp tests/fonts/ahem.ttf ~/.local/share/fonts/ahem.ttf
```
I could add a method on the test setup to copy fonts as needed. Do you agree with this approach @freakboy3742 ?

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,109 @@
+def _check_font_family_linux(value):
+    """List available font family names on linux."""
+    import gi  # noqa
+    gi.require_version("Gtk", "3.0")
```
This was what I found on google using pango. I guess this is more in line with what you had in mind?

goanpeca commented:
>The new font identification stuff is great; however, as I've flagged inline, we proably want to reverse the lookup (i.e., investigate if a font name is valid on request, rather than pre-computing a list of all valid fonts). It also adds some new testing requirements that we need to make sure we've got in CI.

I have added an automatic way to add the needed fonts so users do not really need to follow any special instruction and I added a `FontDatabase` class that checks for fonts on demand and caches them if found so I think this solves your initial concerns 

>Other than that, the big requirement is more testing. The font parsing in particular is a complex piece; I'd expect to see a successful and failing parse case for each font attribute in each possible position, as well as a comprehensive set of cases for the comma-splitting part of the font name parsing.

Added more tests for all the cases you suggested.

freakboy3742 commented on tests/test_parser.py:
```diff
@@ -187,3 +190,554 @@ def test_named_color(self):
+
+        return font_dict
+
+    def test_parse_font_shorthand(self):
```
I've been slowly coming around on pytest, in particular because it makes this kind of parameterized test much cleaner (especially in test output reporting). If you're pytest-inclined, and want to move this to a set of parametrized pytest cases, I'd be OK with that.

goanpeca replied:
Sure, I can update this

freakboy3742 commented on tests/utils.py:
```diff
@@ -125,34 +132,99 @@ def clean_layout(layout):
+        return ('  ' * depth + "* '{text}'\n".format(text=layout['text'].strip()))
+
+
+def copy_fonts(system=False):
```
Although it does do some copying, this is more of an "install" fonts - you don't have any control over the destination.

goanpeca replied:
True, will rename!

freakboy3742 commented on tests/utils.py:
```diff
@@ -125,34 +132,99 @@ def clean_layout(layout):
+            self.copy_fonts()
+
+    def copy_fonts(self):
+        copy_fonts()
```
Any reason for this "copy_fonts calls copy fonts" indirection?

goanpeca replied:
I am using copy_fonts (the func) also on the `if __init__ ...` (for CI)

goanpeca commented on colosseum/parser.py:
```diff
@@ -126,3 +128,133 @@ def color(value):
+    """Parse font shorthand property part for known properties."""
+    from .constants import (FONT_SIZE_CHOICES, FONT_STYLE_CHOICES, FONT_VARIANT_CHOICES,
+                            FONT_WEIGHT_CHOICES, LINE_HEIGHT_CHOICES, NORMAL)
+    font_dict = font_dict.copy()
```
I don't follow you here. Could you elaborate?

freakboy3742 replied:
So - this function accepts a full font dictionary as an argument, takes a copy, and then returns the copy. I'm saying that the function doesn't need to accept the dictionary, just the value; it can parse that value, and return the dictionary of *new* values, the calling function can then merge into the "master" definition as required (and as allowed - since it's possible you could have conflicting definitions, like the "bigger smaller Helvetica" example that I gave)

goanpeca replied:
Gotcha

goanpeca commented on colosseum/units.py:
```diff
@@ -18,10 +18,14 @@ def __init__(self, suffix, val=None):
 
     def __repr__(self):
-        return '{}{}'.format(self.val, self.suffix)
+        int_value = int(self.val)
```
Hmm the repr of '9px' was returning `9.0px` which seemed odd. Is there a case where self.val will not be numeric (and is a valid thing?)

goanpeca replied:
This is the PixelUnit definition so the value should always be numeric AFAIK

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,141 @@
+    NSFontManager.declare_class_property('sharedFontManager')
+    NSFontManager.declare_property("availableFontFamilies")
+    manager = NSFontManager.sharedFontManager
+    for item in manager.availableFontFamilies:
```
I can investigate

goanpeca replied:
this might work https://developer.apple.com/documentation/appkit/nsfont/1525977-fontwithname?language=objc

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,141 @@
+        def check_system_font(self, value):
+            """Check if font family exists on system."""
+            context = self.create_pango_context()
+            for font_family in context.list_families():
```
I can investigate

goanpeca replied:
This might work https://developer.gnome.org/pygtk/stable/class-pangocontext.html#method-pangocontext--load-font

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,141 @@
+                             r"Software\Microsoft\Windows NT\CurrentVersion\Fonts",
+                             0,
+                             winreg.KEY_READ)
+        for idx in range(0, winreg.QueryInfoKey(key)[1]):
```
Not possible AFAICT

https://docs.python.org/3/library/winreg.html

goanpeca replied:
My bad, it is indeed possible, will make the fixes!

goanpeca commented on colosseum/validators.py:
```diff
@@ -97,9 +100,66 @@ def is_color(value):
+
+    This validator returns a list.
+    """
+    from .constants import GENERIC_FAMILY_FONTS as generic_family
```
Circular imports?

freakboy3742 replied:
Ah - right. That's often an indicator of a bigger problem, but in this case, I'm not sure there's going to be a cleaner way to separate things so that we don't have the circular import problem.

goanpeca replied:
We could split choices and constants so that choices holds the Choices logic plus the construction of different choices items and leave constants to only hold constants?

This might now get rid of all issues, but might better decouple the code base.

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,141 @@
+    @staticmethod
+    def fonts_path(system=False):
+        """Return the path for cross platform user fonts."""
+        if os.name == 'nt':
```
I guess I have always used `os.name == 'nt'`... so 🤷‍♂. Can change!

goanpeca replied:
Fixed

goanpeca commented on .github/workflows/ci.yml:
```diff
@@ -25,7 +25,7 @@ jobs:
         beefore --username github-actions --repository ${{ github.repository }} --pull-request ${{ github.event.number }} --commit ${{ github.event.pull_request.head.sha }} ${{ matrix.task }} .
 
   smoke:
-    name: Smoke test
```
Will do

goanpeca replied:
Sounds good 👍

goanpeca replied:
For build status should we test all platforms as well?

goanpeca commented on colosseum/parser.py:
```diff
@@ -126,3 +128,133 @@ def color(value):
+# Font handling
+##############################################################################
+def _parse_font_property_part(value, font_dict):
+    """Parse font shorthand property part for known properties."""
```
👍

goanpeca commented on colosseum/parser.py:
```diff
@@ -126,3 +128,133 @@ def color(value):
+        for _ in range(5):
+            value = parts[0]
+            try:
+                font_dict, is_font_size = _parse_font_property_part(value, font_dict)
```
Hmmm interesting. I think this will 💥 😆

goanpeca replied:
>What happens here in the case of "normal"? _parse_font_property_part ignores normal, but I can't see how that property is picked up if it's manually specified for style or variant?

I skip `normal` during the parsing process because it brings no information regarding which property it refers to: is it a font_style? is it a font_variant? is it a font_weight?

So values that do provide information on these are used. Now since we start with a dictionary of initial values (where all those properties are in facto normal) if the user used normal, then that will be reflected at the end anyway.

>Also, what happens if you specify something daft like "bigger smaller Helvetica" - i.e., double specify one of properties with entirely legal, but contradictory values?

Added a check for this! nice catch

goanpeca commented on colosseum/declaration.py:
```diff
@@ -504,13 +607,25 @@ def keys(self):
-                pass
+            if name == 'font':
+                try:
+                    if getattr(self, '_%s' % name, None):
```
Added two wrappers, `FontShorthand` and `FontFamily`, they do not check values since that will be done when setting the actual values on the style.

goanpeca commented:
@freakboy3742 added the corrections and updted CI to use pytest

goanpeca commented:
I gave it a bit of thought and I don't know if it is worth it to use pytest. I could just use permutations  instead of generating over a thousand different tests. Instructions become less simple and pytest takes longer to run. Thoughts @freakboy3742 ?

goanpeca commented on colosseum/declaration.py:
```diff
@@ -503,14 +610,28 @@ def keys(self):
+                try:
+                    if getattr(self, '_%s' % name, None):
+                        non_default.append((name, getattr(self, name)))
+                except AttributeError:
```
Hmm nope, you are correct.

goanpeca replied:
Fixed (uploading commit soon)

goanpeca commented on colosseum/fonts.py:
```diff
@@ -0,0 +1,140 @@
+    @staticmethod
+    def fonts_path(system=False):
+        """Return the path for cross platform user fonts."""
+        if sys.platform == 'win32':
```
Ok. Could you point me to an example :-) ?

goanpeca replied:
Fixed (uploading commit soon)

goanpeca commented on colosseum/declaration.py:
```diff
@@ -1,33 +1,115 @@
+
+        if values != getattr(self, '_%s' % name, list(initial)):
+            # We use the initial type to make an instance with the set values
+            setattr(self, '_%s' % name, initial.__class__(values))
```
Ok, sounds good

goanpeca replied:
Fixed (uploading commit soon)

goanpeca commented on tests/test_wrappers.py:
```diff
@@ -0,0 +1,148 @@
+        self.assertNotEqual(id(ilist1), id(ilist1.copy()))
+        self.assertNotEqual(id(ilist2), id(ilist1.copy()))
+        self.assertEqual(hash(ilist2), hash(ilist1.copy()))
+        self.assertEqual(ilist1, ilist1.copy())
```
Ok

goanpeca replied:
Fixed (uploading commit soon)

goanpeca replied:
Will add more shorthand tests as soon as I have more clarity on the approach

goanpeca commented on tests/test_wrappers.py:
```diff
@@ -0,0 +1,148 @@
+    def test_font_shorthand(self):
+        # Check initial
+        font = FontShorthand()
+        self.assertEqual(str(font), 'normal normal normal medium/normal initial')
```
Fair enough :-)

goanpeca replied:
@freakboy3742 I kinda disagree because setting the font shorthand to say just `'initial'`, will reset all other font properties to their default values so this is like effectively setting those values to their initial values "implicitly"

I am not sure how we could really handle this use case.

I could try to display the minimum values in this str but once this "Font" object is applied to the style of a given node it is effectively resetting all the other properties.

goanpeca replied:
Still need to think this a bit further

goanpeca commented:
>More solid improvements; still a bit more to go.

>In particular, the FontShorthand handling (and, more generally, the approach we take for "shorthand" properties) needs some work

>It strikes me that we've got two different purposes for FontShorthand. There's a storage class that is "a font" (i.e., this is "italic bold 14pt Helvetica") that can be easily mapped to an actual platform font instance; and there's the validation and storage mechanism for the individual configuration points that make up that definition. Asking for "the font" should give you a single object; however, that doesn't necessarily mean that's how it should be stored. This becomes most apparent in the tests you've added for FontShorthand.__repr__ - a font declaration of "normal normal normal..." ... isn't normal. :-)

I think `FontShorthand` and the `Font` object that should provide whatever font specific properties (em values x height etc...) we might need might be a single object. They could be one and the same I guess.

>I'm wondering if the best approach here might be to make the underlying base properties (e.g., font_family) the storage point for each individual property; with the "shorthand" properties able to either compose the underlying properties into a string representation, or generate an "object" representation as needed.

Not sure, your call :-) 

>Regarding adopting pytest - my general preference is to have more, finer granularity tests. When you have monolithic tests, seeing a test failure doesn't tell you how widespread the problem actually is. The worst outcome for me is when a test fails, you make a fix, and the same test fails, but three lines later. For me, that means you've got 2 different tests; it's vastly preferable to have 2 failures in that situation. Having 1000 tests isn't a problem in my book. If it's causing a problem with test runtime, I'd be interested to know if the pytest-xdist extension helps.

Will look into pytest-xdist

>Also - could I ask that you don't click mark as resolved if it's a question/issue that I've raised? On a long-running ticket like this one, it's difficult for me to tell which pieces of feedback you've addressed and which ones you haven't.

Will do!

goanpeca commented on colosseum/declaration.py:
```diff
@@ -1,33 +1,115 @@
+        except ValueError:
+            raise ValueError("Invalid value '%s' for CSS property '%s'; Valid values are: %s" % (
+                value, name, choices
+            ))
```
Fixed (uploading commit soon)

goanpeca commented on colosseum/declaration.py:
```diff
@@ -503,14 +610,28 @@ def keys(self):
+            # Shorthand values are constructed on demand since they depend on
+            # other properties, we have to use getattr(self, name) instead of
+            # getattr(self, '_%s' % name)
+            if name in shorthand_properties:
```
Fixed (uploading commit soon)

goanpeca commented on colosseum/declaration.py:
```diff
@@ -1,33 +1,108 @@
+
+    def setter(self, value):
+        try:
+            shorthand = storage_class(**parser(value))
```
@freakboy3742 what should be the behavior when setting a font shorthand?

If the user only provides some parts, let's say"

style.font = "12px/3 serif"

Since this shorthand resets the missing properties to their initial values, then should we explicitly write the properties or delete them (which would be akin to setting them to their initial values when the user requests them)

So:

Set properties with missing values
```
style.font = "12px/3 serif"
print(style)
font: ..
font_size: ...
font_family: ...
font_style: ...
font_variant: ...
font_weight: ...
```
vs.

Delete properties with missing values
```
style.font = "12px/3 serif"
print(style)
font: ..
font_size: ...
font_family: ...
```

freakboy3742 replied:
I'd say the latter. "normal normal normal normal 12px/3 serif" doesn't communicate anything that "12px/3 serif" doesn't communicate; and I'd say the same is also true for the inverse.  If I explicitly ask for "what is the font variant", then the answer is "normal"; but if I just ask for the font, the explicitness of including "normal" for font variant doesn't really aid any understanding.

goanpeca replied:
👍
